README.md
# password_hash_plus

[![pub package](https://img.shields.io/pub/v/password_hash_plus)](https://pub.dev/packages/password_hash_plus)

> This plugin is a null-safety version of [password_hash](https://pub.dev/packages/password_hash), you can use it if you
> need Flutter 2.x and above.

Implements PBKDF2 algorithm for securely hashing passwords.

## Preparing for use

### Version constraints

```yaml
  sdk: ">=2.17.0 <3.0.0"
  flutter: ">=3.0.0"
```

## Usage:

```dart
var generator = new PBKDF2();
var salt = Salt.generateAsBase64String();
var hash = generator.generateKey("mytopsecretpassword", salt, 1000, 32);
```


CHANGELOG.md
## 4.0.0

* Migrate to Dart 2.17.0, drop supports for previous Dart versions.
* Migrate to Flutter 3, drop supports for previous Flutter versions.

## 3.0.3

* Add `export 'package:crypto/crypto.dart';` to `lib/password_hash_plus.dart`

## 3.0.2

* Add more detail to the `description` field of `pubspec.yaml`.
* Improve score.

## 3.0.1

* Realize function.

## 3.0.0

* Migrate to non-nullable by default.


test/password_hash_plus_test.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:password_hash_plus/password_hash_plus.dart';
import 'package:test/test.dart';

List<int> encodeBytes(String bytes) {
  final List<String> byteList = bytes.split(' ');
  final List<int> result = <int>[];
  for (final String byte in byteList) {
    result.add(int.parse(byte, radix: 16));
  }
  return result;
}

void main() {
  group('Salt', () {
    test('Can generate random list of integers', () {
      final List<int> salt = Salt.generate(64);
      expect(salt.length, 64);

      expect(salt, everyElement(lessThan(256)));
      expect(salt, everyElement(greaterThanOrEqualTo(0)));
    });

    test('Can generate base64 salt', () {
      final String salt = Salt.generateAsBase64String(64);
      final Uint8List decoded = const Base64Decoder().convert(salt);
      expect(decoded.length, 64);

      expect(decoded, everyElement(lessThan(256)));
      expect(decoded, everyElement(greaterThanOrEqualTo(0)));
    });
  });

  group('RFC6070', () {
    test('Disallow large values of derived key length', () {
      const Hash sha = sha1;
      final int hLen = sha.blockSize;
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha);

      try {
        gen.generateKey('password', 'salt', 1, ((2 << 31) - 1) * hLen + 1);
        expect(true, false);
      } on PBKDF2Exception catch (e) {
        expect(e.toString(), contains('Derived key too long'));
      }
    });

    test('Test vectors 1', () {
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha1);
      final List<int> output = gen.generateKey('password', 'salt', 1, 20);
      expect(
        output,
        encodeBytes(
          '0c 60 c8 0f 96 1f 0e 71 f3 a9 b5 24 af 60 12 06 2f e0 37 a6',
        ),
      );
    });

    test('Test vectors 2', () {
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha1);
      final List<int> output = gen.generateKey('password', 'salt', 2, 20);
      expect(
        output,
        encodeBytes(
          'ea 6c 01 4d c7 2d 6f 8c cd 1e d9 2a ce 1d 41 f0 d8 de 89 57',
        ),
      );
    });

    test('Test vectors 3', () {
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha1);
      final List<int> output = gen.generateKey('password', 'salt', 4096, 20);
      expect(
        output,
        encodeBytes(
          '4b 00 79 01 b7 65 48 9a be ad 49 d9 26 f7 21 d0 65 a4 29 c1',
        ),
      );
    });

    // This test may take a few minutes to run
    test('Test vectors 4', () {
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha1);
      final List<int> output =
          gen.generateKey('password', 'salt', 16777216, 20);
      expect(
        output,
        encodeBytes(
          'ee fe 3d 61 cd 4d a4 e4 e9 94 5b 3d 6b a2 15 8c 26 34 e9 84',
        ),
      );
    });

    test('Test vectors 5', () {
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha1);
      final List<int> output = gen.generateKey(
        'passwordPASSWORDpassword',
        'saltSALTsaltSALTsaltSALTsaltSALTsalt',
        4096,
        25,
      );
      expect(
        output,
        encodeBytes(
          '3d 2e ec 4f e4 1c 84 9b 80 c8 d8 36 62 c0 e4 4a 8b 29 1a 96 4c f2 '
          'f0 70 38',
        ),
      );
    });

    test('Test vectors 6', () {
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha1);
      final List<int> output =
          gen.generateKey('pass\u0000word', 'sa\u0000lt', 4096, 16);
      expect(
        output,
        encodeBytes('56 fa 6a a7 55 48 09 9d cc 37 d7 f0 34 25 e0 c3'),
      );
    });
  });

  group('Sha256', () {
    test('Disallow large values of derived key length', () {
      const Hash sha = sha256;
      final int hLen = sha.blockSize;
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha);

      try {
        gen.generateKey('password', 'salt', 1, ((2 << 31) - 1) * hLen + 1);
        expect(true, false);
      } on PBKDF2Exception catch (e) {
        expect(e.toString(), contains('Derived key too long'));
      }
    });

    test('Test vectors 1', () {
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha256);
      final List<int> output = gen.generateKey('password', 'salt', 1, 32);
      expect(
        output,
        encodeBytes(
          '12 0f b6 cf fc f8 b3 2c 43 e7 22 52 56 c4 f8 37 a8 65 48 c9 2c cc '
          '35 48 08 05 98 7c b7 0b e1 7b',
        ),
      );
    });

    test('Test vectors 2', () {
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha256);
      final List<int> output = gen.generateKey('password', 'salt', 2, 32);
      expect(
        output,
        encodeBytes(
          'ae 4d 0c 95 af 6b 46 d3 2d 0a df f9 28 f0 6d d0 2a 30 3f 8e f3 c2 '
          '51 df d6 e2 d8 5a 95 47 4c 43',
        ),
      );
    });

    test('Test vectors 3', () {
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha256);
      final List<int> output = gen.generateKey('password', 'salt', 4096, 32);
      expect(
        output,
        encodeBytes(
          'c5 e4 78 d5 92 88 c8 41 aa 53 0d b6 84 5c 4c 8d 96 28 93 a0 01 ce '
          '4e 11 a4 96 38 73 aa 98 13 4a',
        ),
      );
    });

    // This test may take a few minutes to run
    test('Test vectors 4', () {
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha256);
      final List<int> output =
          gen.generateKey('password', 'salt', 16777216, 32);
      expect(
        output,
        encodeBytes(
          'cf 81 c6 6f e8 cf c0 4d 1f 31 ec b6 5d ab 40 89 f7 f1 79 e8 9b 3b '
          '0b cb 17 ad 10 e3 ac 6e ba 46',
        ),
      );
    });

    test('Test vectors 5', () {
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha256);
      final List<int> output = gen.generateKey(
        'passwordPASSWORDpassword',
        'saltSALTsaltSALTsaltSALTsaltSALTsalt',
        4096,
        40,
      );
      expect(
        output,
        encodeBytes(
          '34 8c 89 db cb d3 2b 2f 32 d8 14 b8 11 6e 84 cf 2b 17 34 7e bc 18 '
          '00 18 1c 4e 2a 1f b8 dd 53 e1 c6 35 51 8c 7d ac 47 e9',
        ),
      );
    });

    test('Test vectors 6', () {
      final PBKDF2 gen = PBKDF2(hashAlgorithm: sha256);
      final List<int> output =
          gen.generateKey('pass\u0000word', 'sa\u0000lt', 4096, 16);
      expect(
        output,
        encodeBytes('89 b6 9d 05 16 f8 29 89 3c 69 62 26 65 0a 86 87'),
      );
    });
  });
}


lib/src/salt.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

/// @Describe: Salt
///
/// @Author: LiWeNHuI
/// @Date: 2022/5/1

/// Generates salts for hashing algorithms.
class Salt {
  /// Generates a random salt of [length] bytes from a cryptographically secure
  /// random number generator.
  ///
  /// Each element of this list is a byte.
  static List<int> generate(int length) {
    final Uint8List buffer = Uint8List(length);
    final Random rng = Random.secure();
    for (int i = 0; i < length; i++) {
      buffer[i] = rng.nextInt(256);
    }
    return buffer;
  }

  /// Generates a random salt of [length] bytes from a cryptographically secure
  /// random number generator and encodes it to Base64.
  ///
  /// [length] is the number of bytes generated, not the [length] of the base64
  /// encoded string returned. Decoding
  ///
  /// the base64 encoded string will yield [length] number of bytes.
  static String generateAsBase64String(int length) {
    const Base64Encoder encoder = Base64Encoder();
    return encoder.convert(generate(length));
  }
}


lib/src/pbkdf2.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

import 'salt.dart';

/// @Describe: PBKDF2
///
/// @Author: LiWeNHuI
/// @Date: 2022/5/1

/// Instances of this type derive a key from a password, salt, and hash function
///
/// https://en.wikipedia.org/wiki/PBKDF2
class PBKDF2 {
  /// Creates instance capable of generating a key.
  ///
  /// [hashAlgorithm] defaults to [sha256].
  PBKDF2({Hash? hashAlgorithm}) {
    this.hashAlgorithm = hashAlgorithm ?? sha256;
  }

  Hash get hashAlgorithm => _hashAlgorithm;

  set hashAlgorithm(Hash algorithm) {
    _hashAlgorithm = algorithm;
    _blockSize = _hashAlgorithm.convert(<int>[1, 2, 3]).bytes.length;
  }

  late Hash _hashAlgorithm;
  late int _blockSize;

  /// Hashes a [password] with a given [salt].
  ///
  /// The length of this return value will be [keyLength].
  ///
  /// See [Salt.generateAsBase64String] for generating a random salt.
  ///
  /// See also [generateBase64Key], which base64 encodes the key returned from
  /// this method for storage.
  List<int> generateKey(
    String password,
    String salt,
    int rounds,
    int keyLength,
  ) {
    if (keyLength > (pow(2, 32) - 1) * _blockSize) {
      throw PBKDF2Exception('Derived key too long');
    }

    final int numberOfBlocks = (keyLength / _blockSize).ceil();
    final Hmac hmac = Hmac(hashAlgorithm, utf8.encode(password));
    final ByteData key = ByteData(keyLength);
    int offset = 0;

    final List<int> saltBytes = utf8.encode(salt);
    final int saltLength = saltBytes.length;
    final ByteData inputBuffer = ByteData(saltBytes.length + 4)
      ..buffer.asUint8List().setRange(0, saltBytes.length, saltBytes);

    for (int blockNumber = 1; blockNumber <= numberOfBlocks; blockNumber++) {
      inputBuffer
        ..setUint8(saltLength, blockNumber >> 24)
        ..setUint8(saltLength + 1, blockNumber >> 16)
        ..setUint8(saltLength + 2, blockNumber >> 8)
        ..setUint8(saltLength + 3, blockNumber);

      final Uint8List block =
          _XORDigestSink.generate(inputBuffer, hmac, rounds);
      int blockLength = _blockSize;
      if (offset + blockLength > keyLength) {
        blockLength = keyLength - offset;
      }
      key.buffer.asUint8List().setRange(offset, offset + blockLength, block);

      offset += blockLength;
    }

    return key.buffer.asUint8List();
  }

  /// Hashed a [password] with a given [salt] and base64 encodes the result.
  ///
  /// This method invokes [generateKey] and base64 encodes the result.
  String generateBase64Key(
    String password,
    String salt,
    int rounds,
    int keyLength,
  ) {
    const Base64Encoder converter = Base64Encoder();

    return converter.convert(generateKey(password, salt, rounds, keyLength));
  }
}

/// Thrown when [PBKDF2] throws an exception.
class PBKDF2Exception implements Exception {
  PBKDF2Exception(this.message);

  String message;

  @override
  String toString() => 'PBKDF2Exception: $message';
}

class _XORDigestSink extends Sink<Digest> {
  _XORDigestSink(ByteData inputBuffer, Hmac hmac) {
    lastDigest = Uint8List.fromList(
      hmac.convert(inputBuffer.buffer.asUint8List()).bytes,
    );
    bytes = ByteData(lastDigest.length)
      ..buffer.asUint8List().setRange(0, lastDigest.length, lastDigest);
  }

  static Uint8List generate(ByteData inputBuffer, Hmac hmac, int rounds) {
    final _XORDigestSink hashSink = _XORDigestSink(inputBuffer, hmac);

    // If rounds == 1, we have already run the first hash in the constructor
    // so this loop won't run.
    for (int round = 1; round < rounds; round++) {
      hmac.startChunkedConversion(hashSink)
        ..add(hashSink.lastDigest)
        ..close();
    }

    return hashSink.bytes.buffer.asUint8List();
  }

  late ByteData bytes;
  late Uint8List lastDigest;

  @override
  void add(Digest digest) {
    lastDigest = Uint8List.fromList(digest.bytes);
    for (int i = 0; i < digest.bytes.length; i++) {
      bytes.setUint8(i, bytes.getUint8(i) ^ lastDigest[i]);
    }
  }

  @override
  void close() {}
}
